### PR TITLE
automatically append newline chars in kprintf when appropriate

### PR DESCRIPTION
**Pull request by GhostLyrics**
+ only when there isn't one already at the last place in array
+ only when it is not part of context coloring for debug

### CODE REVIEW

File: common/source/console/kprintf.cpp [hunk 1]
```diff
@@ -87,6 +87,25 @@ void kprintfd(const char *fmt, ...)
...
+  // and the string is not part of context coloring
+  if (fmt[string_size - 1] == '\n'
+      || !strncmp(fmt, DEBUG_FORMAT_STRING, string_size))
+  {
+    kvprintf(fmt, kprintfd_func, 0, 10, args);
+  }
+  else
+  {
+    kvprintf(fmt, kprintfd_func, 0, 10, args);
+    kvprintf("\n", kprintfd_func, 0, 10, args);
```
Comment (GhostLyrics) [inline common/source/console/kprintf.cpp]: Unfortunately I'm not sure if the after `kprintfd_func` are okay for this. Ideally I'd supply an empty `args` with this, no?

File: common/source/console/kprintf.cpp [hunk 2]
```diff
@@ -87,6 +87,25 @@ void kprintfd(const char *fmt, ...)
   va_list args;
 
   va_start(args, fmt);
-  kvprintf(fmt, kprintfd_func, 0, 10, args);
+
+  int string_size = 0;
```
Comment (GhostLyrics) [inline common/source/console/kprintf.cpp]: After some hours of sleep, I figure this should be solved with `strlen()`

File: common/source/console/kprintf.cpp [hunk 3]
```diff
@@ -87,6 +87,25 @@ void kprintfd(const char *fmt, ...)
...
+  while(fmt[string_size] != '\0')
+  {
+    string_size++;
+  }
+
+  // only append a newline character if there isn't one already
+  // and the string is not part of context coloring
+  if (fmt[string_size - 1] == '\n'
```
Comment (skogler) [inline common/source/console/kprintf.cpp]: Can the string have a size of 0? If so, I'd check for that here.

**Comment by GhostLyrics** [comment]
Updated according to feedback. Still unsure about additional parameters though.

**Comment by dgruss** [comment]
I'm not sure about that... seems pretty non-standard and kprintfd is used a lot to put together debug output...

**Comment by GhostLyrics** [comment]
I argued for a more convenient `debug` because as student you always forget the newline when doing rapid testing, which seemed to make the most sense in this way, since `debug` is a macro. 

I've made sure that nothing gets screwed up by the color output by using the constant which changes based on `NOCOLOR`.

If you've ideas for improvement I'm all ears.

**Comment by GhostLyrics** [comment]
When enabling all debug flags, I've come across a single problem which I'll have another look at. That is the bitmap output in `common/source/util/Bitmap: Bitmap::bmprint` which is printed with every bit in a single line.

**Comment by dgruss** [comment]
then i would rather add it to debug, not to kprintfd... kprintfd must work for cases such as bmprint

**Pull request closed without merging**